### PR TITLE
ELPP-2983 - profile page - login control

### PR DESCRIPTION
**Pull request by davidcmoulton**

### CODE REVIEW

**Comment by davidcmoulton** [comment]
When the login control is finalised, the patterns involved in including it in the site header must be updated accordingly

File: .gitignore [hunk 1]
```diff
@@ -20,3 +18,5 @@ test/build
 
 /vendor/
 /test/fixtures-local/
+/.idea/
```
Comment (thewilkybarkid) [inline .gitignore]: I'd keep stuff like this to your global `.gitignore`.

File: source/_patterns/01-molecules/navigation/profile-login-control.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "button": {
+    "text": "Login / Register",
```
Comment (thewilkybarkid) [inline source/_patterns/01-molecules/navigation/profile-login-control.json]: `Log in` (`Login` is a noun).

File: source/_patterns/01-molecules/navigation/profile-login-control.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      minLength: 1
+    logoutUri:
+      type: string
+      minLength: 1
+    displayName:
+      type: string
+      minLength: 1
+    button:
+      $ref: ../../00-atoms/components/button.yaml#/schema
+#   This dependencies block isn't working as expected: validation passes when isLoggedIn but not logoutUri is present
```
Comment (thewilkybarkid) [inline source/_patterns/01-molecules/navigation/profile-login-control.yaml]: `dependencies` is indented too much.

File: source/_patterns/01-molecules/navigation/profile-login-control.mustache [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<div class="profile-login-control" {{#isLoggedIn}}data-profile-home-link="{{profileHomeLink}}"
+     data-profile-manage-link="{{profileManageLink}}" data-logout-uri="{{logoutUri}}" data-display-name="{{displayName}}"{{/isLoggedIn}}
```
Comment (thewilkybarkid) [inline source/_patterns/01-molecules/navigation/profile-login-control.mustache]: Is there anyway we could decouple what appears in the menu from the template? (So we could add a new user-specific page in Journal without having to change the pattern.)

File: source/_patterns/04-pages/article--magazine.json [hunk 5]
```diff
@@ -90,10 +90,21 @@
         {
           "text": "Submit my research",
           "path": "#",
-          "classes": "nav-secondary__item nav-secondary__item--button",
+          "classes": "nav-secondary__item",
```
Comment (thewilkybarkid) [inline source/_patterns/04-pages/article--magazine.json]: Do we need `classes` as a property anymore?

File: test/profilelogincontrol.spec.js [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+        'shouldOnlyContainCharactersUseableWithinJavaScriptNames;plus-hyphens, commas, andSpaces',
+        'shouldOnlyContainCharactersUseableWithinJavaScriptNames:plus-hyphens, commas, andSpaces',
+        'shouldOnlyContainCharactersUseableWithinJavaScriptNames/plus-hyphens, commas, andSpaces',
+        'hyphens-should-not-be-at-the-end-of-the-string-',
+        'hyphens-should-not-be-at-the-end-of-a-word-, in-the-string',
+        '-hyphens-should-not-be-at-the-start-of-the-string',
+        'hyphens-should-not-be-at-the-start-of-a, -word-in-the-string'
+      ];
+
+      validDataAttributeRoots = [
```
Comment (davidcmoulton) [inline test/profilelogincontrol.spec.js]: Aim to get a regex fitting the final version of these into the json schema, for extra stringency.

**Comment by thewilkybarkid** [comment]
Are we able to break the profile icon part into an image (with CSS rounded borders) and the arrow? The image _could_ be data too, so Journal would set it as the current one, and when we allow avatars there's no pattern change?

File: source/_patterns/04-pages/article--magazine.json (hunk 5 above)
Comment (davidcmoulton) [reply source/_patterns/04-pages/article--magazine.json]: In some cases, yes (see https://github.com/elifesciences/pattern-library/blob/develop/source/_patterns/04-pages/article--magazine.json#L45). The more general case (primary/secondary) might be simplified through refactoring, but I think that's a separate task.

**Comment by davidcmoulton** [comment]
We absolutely could, but I don't think that's MVP for first iteration.

**Comment by davidcmoulton** [comment]
On investigation, it turns out that the image path inside the js file is not being rewritten correctly in https://github.com/elifesciences/journal/blob/develop/gulpfile.js#L139. To work around this, now including icon data as data attributes because, as @thewilkybarkid says, we will need it anyway later.

File: source/_patterns/01-molecules/navigation/login-control.mustache [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+     data-link-field-roots="{{linkFieldRoots}}"
+     {{{linkFieldData}}}
+     data-default-uri="{{defaultUri}}"
+     data-display-name="{{displayName}}"
+
+     {{#icon}}
+       data-icon-path-root="{{pathRoot}}"
+       data-icon-srcset="{{srcset}}"
+       data-icon-alt-text="{{altText}}"
+     {{/icon}}
```
Comment (thewilkybarkid) [inline source/_patterns/01-molecules/navigation/login-control.mustache]: In Carouseltem we embed the `<picture>` (https://github.com/elifesciences/pattern-library/blob/develop/source/_patterns/01-molecules/components/carousel-item.mustache#L3-L19)

Comment (davidcmoulton) [reply source/_patterns/01-molecules/navigation/login-control.mustache]: I've never been totally happy with that approach, but it's the carousel, so let it go. In this case I think it's worth trying to get the data provided by journal and used by pattern-library (have always got the Carouseltem approach to fall back on if we need to).

File: assets/js/components/LoginControl.js [hunk 8]
```diff
@@ -176,12 +176,19 @@ module.exports = class LoginControl {
...
+   * @param {HTMLElement} $elm pattern-root element
    * @param {Function} buildElement Function used to build an element (elife-utils.buildElement)
    * @return {HTMLElement} The resulting control element
    */
-  buildControl(extraLinksToBuild, buildElement) {
+  buildControl(extraLinksToBuild, $elm, buildElement) {
+    const iconData = {
+      pathRoot: $elm.dataset.iconPathRoot,
+      srcset: $elm.dataset.iconSrcset,
+      altText: $elm.dataset.iconAltText
```
Comment (davidcmoulton) [inline assets/js/components/LoginControl.js]: `sizes`

File: source/_patterns/01-molecules/navigation/login-control.mustache (hunk 7 above)
Comment (thewilkybarkid) [reply source/_patterns/01-molecules/navigation/login-control.mustache]: I'm not that big a fan either, but it does allow us to use enable things like WebP (through `<source>`).

File: assets/js/components/LoginControl.js [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+   * @returns {HTMLElement} the <picture> describing the icon
+   */
+  static buildIcon(iconData, buildElement) {
+    if (!(iconData.pathRoot && iconData.srcset && iconData.altText)) {
+      return null;
+    }
+
+    const $icon = buildElement.call(null, 'picture');
+    const $source = buildElement.call(null, 'source', [], '', $icon);
+    $source.setAttribute('srcset', `${iconData.pathRoot}.svg`);
```
Comment (thewilkybarkid) [inline assets/js/components/LoginControl.js]: Journal should just provide a complete URI. User-loaded avatars won't (probably) be SVG.

File: source/_patterns/01-molecules/navigation/login-control.mustache (hunk 7 above)
Comment (davidcmoulton) [reply source/_patterns/01-molecules/navigation/login-control.mustache]: Good point. Okay, will go with the CarouselItem approach and see how it plays.

**Pull request merged**